src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,111 @@
+            }
+
+            public override AnalysisResult VisitSwitch(ISwitchOperation operation, object unused)
+            {
+                // Fail if the switch statement is empty or any of sections have more than one `case` label
+                // Once we have "or" patterns, we can relax this to accept multi-case sections.
+                if (operation.Cases.Length == 0 || operation.Cases.Any(@case => @case.Clauses.Length != 1))
+                {
+                    return AnalysisResult.Failure;
+                }
+
+                // Iterate over all sections and match section bodies to
+                // see if they are all well-formed for a switch expression
+                return Aggregate(operation.Cases, (result, @case) => AnalysisResult.Match(result, VisitList(@case.Body)));
+            }
+
+            private AnalysisResult VisitList(ImmutableArray<IOperation> operations)
+            {
+                // This is eaither a block or switch section body. Here we "combine" the result, since there could be
+                // compatible statements like the ending `break;` or multiple simple assignments
+                return Aggregate(operations, (result, operation) => AnalysisResult.Combine(result, Visit(operation, /*unused*/argument: null)));
+            }
+
+            private static AnalysisResult Aggregate<T>(ImmutableArray<T> operations, Func<AnalysisResult, T, AnalysisResult> func)
+            {
+                var result = AnalysisResult.Neutral;
+                foreach (var operation in operations)
+                {
+                    result = func(result, operation);
+                    if (result.IsFailure)
+                    {
+                        // No point to continue if any operation is
+                        // not well-formed for a switch expression 
+                        break;
+                    }
+                }
+
+                return result;
+            }
+
+            public override AnalysisResult VisitReturn(IReturnOperation operation, object unused)
+            {
+                return AnalysisResult.Return;
+            }
+
+            public override AnalysisResult VisitSimpleAssignment(ISimpleAssignmentOperation operation, object unused)
+            {
+                // Visit the target which could a local or field reference,
+                // in which case we can assign the switch expression to it.
+                return Visit(operation.Target, /*unused*/argument: null);
+            }
+
+            public override AnalysisResult VisitLocalReference(ILocalReferenceOperation operation, object unused)
+            {
+                return AnalysisResult.Assignment(operation.Local);
+            }
+
+            public override AnalysisResult VisitFieldReference(IFieldReferenceOperation operation, object unused)
+            {
+                return AnalysisResult.Assignment(operation.Field);
+            }
```

Comment:
this also feels limiting because you might have something like:


```c#
case X:
   a.b.c = 1;
   break;
case Y:
    a.b.c = 2;
    break;
```

In general, our features detect and understand this sort of thing.  For example "use object initializer' properly matches against this sort of thing.